### ISSUE TITLE
Clean Urls

### ISSUE THREAD

**Issue by gtbu** [cmb69/pfw_xh#29]
In the forum i have read about an api in this plugin which simply enables clean urls - when will this come ?
The other question ist why it is not possible to publish a XH - Version with clean urls as it exists up to 1.6.10 - or is this in the plans for the next version

**Comment by cmb69** [comment]
> In the forum i have read about an api in this plugin which simply enables clean urls - when will this come ?

A most basic implementation is already there, see [class Url](https://github.com/cmb69/pfw_xh/blob/0.2.0/classes/Url.php). 

> The other question ist why it is not possible to publish a XH - Version with clean urls as it exists up to 1.6.10 - or is this in the plans for the next version

The issue is not so much that it is not possible to support clean URLs, but rather that almost no plugin caters to that.

**Comment by gtbu** [comment]
Well - on http://net.square7.ch/XH1610/XH1610clean-0.5/Download is a version 1.6.10 which functions (with slight problems) with almost all plugins. Is the principle not applicable for 1.72 ?

**Comment by cmb69** [comment]
The problem are plugins which construct "normal" URLs; these will still work, but they're creating duplicate content (DC), which is bad for SEO purposes, and as such (at least partially) defeat the purpose of clean URLs.

**Comment by gtbu** [comment]
I think it should be possible to redirect double URLs in the htaccess (and propose to publish a clean beta - at least in the forum)

**Comment by cmb69** [comment]
> I think it should be possible to redirect double URLs in the htaccess

No, you can't – at least with the current clean URL mods, since they rewrite the clean URLs to "classic" URLs.

**Comment by gtbu** [comment]
Can You give me an example of a clean url and of the double url of Cmsimple XH. One can remove the ? by a dom-model-code but that looses the pages.

**Comment by gtbu** [comment]
Here is a url mit an example : function make301($goto) 

https://www.abakus-internet-marketing.de/foren/viewtopic.php?t=98261&highlight=fragezeichen++url+entfernen

which can be adapted

**Comment by cmb69** [comment]
And here is an already working solution for CMSimple_XH: https://github.com/cmb69/cmsimple-xh/tree/clean-urls

**Comment by gtbu** [comment]
Sorry - on xampp 7.025 the clean-url-master has no clean urls ...

**Comment by cmb69** [comment]
> also to rename the _htaccess to .htaccess has no effect -

Did you _also_ enable the config option Uri → Clean?

**Comment by gtbu** [comment]
I can oly find URI  :  1. Separator and 2.Word separator

**Comment by cmb69** [comment]
> I can only find URI : 1. Separator and 2.Word separator

Oh, sorry. I've made that a hidden setting, so you'd have to change [this line](https://github.com/cmb69/cmsimple-xh/blob/clean-urls/cmsimple/config.php#L44) to:
````
$cf['uri']['clean']="true";
````

**Comment by gtbu** [comment]
That helps only partially - only for start - languages and menulevels. The others give errors
Can You please change the master so that it functions

**Comment by gtbu** [comment]
As i see exist these problems only on Xampp. But this clean version has still the double /?sitename and /sitename , even though google will hardly scan /?sitename but /index.php?...

**Comment by gtbu** [comment]
As i found in the clean version : The cms hangs a ? behind the links of previous an next (not #Top) in the footer - though it functions

**Comment by cmb69** [comment]
> As i found in the clean version (not at 3-magi.net) : The cms hangs a ? behind the links of previous and next (not #Top) in the footer and in the url - though it functions

Thanks! Indeed, I'd stumbled over the misuse of the second parameter of `a()` again. It appears to be best to get rid of this misuse, generally.

**Comment by cmb69** [comment]
> That helps only partially - only for start - languages and menulevels. The others give errors

There are two unrelated issues. Firstly, the clean URL solution doesn't work well with some special characters, which is already tracked by cmsimple-xh/cmsimple-xh#357. Secondly, the clean URL solution won't be able to allow page names which are identical to existing files and folders (here, `templates` and `plugins`). Not that much of a problem on case-sensitive file systems, but a bigger issue for case-insensitive file systems such as NTFS.
